### DISCUSSION TITLE
Using Core0

### DISCUSSION THREAD

**Discussion by MichaelPNolan** [marcel-licence/esp32_basic_synth#58] [General]
I moved some processing to the Core0 task loop

At first it wasn't being called but I checked your code and it was updated to the MIDI_VIA_USB... module
so adding that in now it works.
```
#if (defined ADC_TO_MIDI_ENABLED) || (defined MIDI_VIA_USB_ENABLED)
    xTaskCreatePinnedToCore(Core0Task, "Core0Task", 8000, NULL, 999, &Core0TaskHnd, 0);
#endif
```

In the first 30-60 seconds I get audio glitches - even one where it was making a rough popping on left side then right side alternating.
So processing Core0 seems to slow auto processing which we assume only happens on Core1 in arduino.
Therefore, generally no improvement - but over time things seem to buffer or perform ok.
So I'll look at having a kind of reduced frequency of calls in the Task0loop  using some kind of loop count variable to see if that removes that glitch.

**Comment by marcel-licence** [comment]
Do you run your own version actually or the latest, original code from GitHub?

**Comment by MichaelPNolan** [comment]
I guess it's my own version at this point. So I'd have to map all my
changes to upgrade. Not that many changes, trying to use per design. Mostly
add ons.

Regards
Michael

**Comment by marcel-licence** [comment]
Hm okay when there are modifications then it would be very hard to support. I would like to convert this issue into a discussion then.

**Comment by MichaelPNolan** [comment]
Ok - you made me realize that. However, i was just surprised since I was
doing it all without Core0Task and then I got ... less performance by using
it?
So that part ... is more like an issue but its not shared software. Was
just thinking you might have tips or have experienced same along the way.
Please convert.
Regards
Michael

**Reply by MichaelPNolan** [reply]
Ok - you can close this. So here is my diagnosis theory.
I set it up like you did - as related to my changes (which are pretty minimal). I believe that if you put the MidiUSB processing in Task0 somehow because of communications SPI it creates timing issues and thus the audio lags and makes audio artifacts.

So, like you, my analogue and button related checking is in the Core0 task loop and the USBMidi etc is back in the main loop. It is still true that I have to call UsbMidi_Loop(); quite infrequently. If it is called to often it also interferes with DAC process timing.

I forked this now ... in case anyone cares. I've not had any repositories on Github but I'll start putting all my things related to this up.

**Comment by marcel-licence** [comment]
Yes I can confirm doing SPI during the audio loop is very problematic. In addition to that there are two functions for USB processing because otherwise you can also get problems when both cores want to access same data or may interfere with each other.

Thank you for your feedback =)
